tensorflow_tts/processor/base_processorv2.py

```diff
@@ -0,0 +1,140 @@
+from dataclasses import dataclass
+import abc
+import os
+import json
+
+from typing import Union
+
+
+@dataclass
+class ExamplePrepro(abc.ABC):
+    root_dir: str
+    symbols: list
+    speakers_map = {}
+    train_f_name: str = "train.txt"
+    delimiter: str = "|"
+    positions = {"file": 0, "text": 1, "speaker_name": 2}  # positions of file,text,speaker_name after split line
+    f_extension: str = ".wav"
+    extra_tokens = {"unk": "[UNK]", "pad": "[PAD]", "eos": "[EOS]", "bos": "[BOS]"}
+    save_mapper: bool = False
+
+    # extras
+    items = []
+    symbol_to_id = {}
+    id_to_symbol = {}
+
+    def __post_init__(self):
+        self.create_items()
+        self.create_speaker_map()
+        self.reverse_speaker = {v: k for k, v in self.speakers_map.items()}
+        self.create_symbols()
+        if self.save_mapper:
+            self.__save_mapper()
+
+    def __getattr__(self, name: str) -> str:
+        if "_id" in name:  # map extra token to id
+            return self.symbol_to_id[self.extra_tokens[name.split("_id")[0]]]
+        return self.extra_tokens[name]  # map extra token to value
+
+    def create_speaker_map(self):
+        """
+        Create speaker map for dataset
+        """
+        sp_id = 0
+        for i in self.items:
+            speaker_name = i[-1]
+            if speaker_name not in self.speakers_map:
+                self.speakers_map[speaker_name] = sp_id
+                sp_id += 1
+
+    def get_speaker_id(self, name: str) -> int:
+        return self.speakers_map[name]
+
+    def get_speaker_name(self, speaker_id: int) -> str:
+        return self.speakers_map[speaker_id]
+
+    def create_symbols(self):
+        self.symbols = self.symbols + list(self.extra_tokens.values())
```

Comment:
Because the backward compatibility, the symbols should [pad] + symbold + [eos] + [bos] + [unk].

Reply:
We can save mapper to json so shouldnt consider makinig random constant positions

Reply:
i mean the extra token. We should make our new processor compatible with old version (https://github.com/TensorSpeech/TensorFlowTTS/blob/master/tensorflow_tts/processor/ljspeech.py#L110-L112).

Reply:
also, can you add pad_id, eos_id, bos_id, unk_id as attribute of the processor. So we can easily check it indices by processor.pad_id, processor.eos_id.

Reply:
Its added as attributes __getattr__ method take care of it :)

Reply:
```python3
@dataclass
class LJ(ExamplePrepro):

    def get_one_sample(self, item):
        sample = {
            "raw_text": None,
            "text_ids": None,
            "audio": None,
            "utt_id": None,
            "speaker_name": None,
            "rate": None,
        }
        return sample

    def create_items(self):
        for i in range(2):
            self.items.append(["MY OWN CREATE ITEM", "MY PATH", "ME!"])

    def text_to_sequence(self, text):
        return ["0"]

e = LJ("files", list(string.ascii_lowercase))
print(e.eos)
print(e.eos_id)
```

Reply:
> i mean the extra token. We should make our new processor compatible with old version (https://github.com/TensorSpeech/TensorFlowTTS/blob/master/tensorflow_tts/processor/ljspeech.py#L110-L112).

Sure but I'm not sure if we should consider new processor being fully compatible with old version as old version wasn't abstract and have a lot of magic positions in it.

We can just add old version (like tensorflowv1 etc :D) to old_preprocess subfolder if we really want to have backwards compatible but right now based on pre 1.0 versions all changes can be breaking (but not all shouldn't :P) 

https://docs.npmjs.com/about-semantic-versioning